pkgs/build-support/fetchurl/builder.sh

```diff
@@ -57,12 +57,15 @@ tryHashedMirrors() {
     fi
 
     for mirror in $hashedMirrors; do
-        url="$mirror/$outputHashAlgo/$outputHash"
+        url="$mirror/$fetchType/$outputHash"
         if $curl --retry 0 --connect-timeout "${NIX_CONNECT_TIMEOUT:-15}" \
             --fail --silent --show-error --head "$url" \
             --write-out "%{http_code}" --output /dev/null > code 2> log; then
             tryDownload "$url"
-            if test -n "$success"; then finish; fi
+            if test -n "$success"; then
+                tar -xxf $downloadedFile -C $(dirname $downloadedFile)
```

Comment:
It's a poor approach to integrate some data compression. With tar the routine in copy-tarballs.pl is quite universal and works independently of whether files or directories are cached. But yes, it's hacky, I'll improve on that, thanks.